---
title: IPGWAS
layout: post
categories: [Bioinformatics]
tags: [GWAS,Perl]
image: /figure
---
{% include JB/setup %}

IPGWAS was developed to facilitate the identification of the rational thresholds in QC of GWAS datasets, association analysis, Manhattan plot, quantile-quantile (QQ) plot, and format conversion for genetic analyses, such as meta-analysis, genotype phasing, and imputation. IPGWAS is a multiplatform application written in Perl with a graphical user interface (GUI).

##### Features
* genome-wide association study (GWAS)
* quality control (qc)
* Manhattan plot and QQ plot
* Cochran-Armitage trend test, Association test
* combine GWAS dataset, Split GWAS Data by Chromosome
* Convert EIGENSTRAT output chi-square to p value
* Convert MACH imputation output to PED/MAP format and SNPTEST format
* Convert PED/MAP files to the default input format of PHASE and BEAGLE
* Convert PLINK and SNPTEST files to GWAMA format
* P-Value Calculator (Chi-square test, Cochran-Armitage trend test, and Fisher's exact test)
* Change affection status of GWAS data
* filt Subjects of GWAS data by affection status (case/control) and/or gender (male/female)
* GUI for PLINK, SNP ratio Test
* filter singleton significant SNPs
* eigPlot (plot the eigenstrat result)
* Convert PED/MAP files to the default input format of MACH and IMPUTE2
* Convert IMPUTE2 imputation output to TPED/TFAM (PLINK) format

IPGWAS is available at [sourceforge](http://sourceforge.net/projects/ipgwas/).
